histogram/src/bucket.rs

```diff
@@ -1,29 +1,33 @@
-// Copyright 2022 Twitter, Inc.
-// Licensed under the Apache License, Version 2.0
-// http://www.apache.org/licenses/LICENSE-2.0
+//! Buckets represent quantized value ranges and a count of observations within
+//! that range.
 
-/// A `Bucket` represents a discrete range of values and the sum of recorded
-/// counts within this range.
-#[derive(Clone, Copy, PartialEq)]
+/// A bucket represents a quantized range of values and a count of observations
+/// that fall into that range.
+#[derive(Copy, Clone, Debug, PartialEq)]
 pub struct Bucket {
-    pub(crate) low: u64,
-    pub(crate) high: u64,
-    pub(crate) count: u32,
+    pub(crate) count: u64,
```

Comment:
Curious about the bump of count from u32 to u64.

Reply:
Mostly about aligning this histogram implementation with the Rezolus BPF programs, which use 64bit counters in kernel space.

I'm open to hearing arguments that we should stick to 32bit counters.

A 0,7,64 histogram uses 7424 buckets and covers the entire 64bit range with a relative error of 0.78% - it would increase to 58KB if we stick to 64bit counters. A 1s resolution, 1m span sliding window histogram would use 62 * 58KB => ~3.5MB. However, we're unlikely to see double-wraps in between snapshots.

If we think it's also unlikely to see double-wraps using 32bit counters, I'm happy to drop it down again.

Reply:
What if the free running histogram has 64-bit counters to match BPF, but we store the sliding window delta (`distribution_since()`) in a 32-bit representation (perhaps with overflow detection/warning)? There's advantage in keeping the free-running histogram on 64-bit counters, because raw value can definitely doubly wrap around 32-bit (e.g. an 200k / sec event generates 17 billion updates each day, and if a bucket receives half of that it'll wrap around a 32-bit counter twice, in this case a "daily delta" will produce the wrong number between two snapshots a day apart). Such sparse snapshots should always use the raw histogram and users can handle the diffing manually. But for any meaningful and typical sliding window intervals (usually on the order of milliseconds, seconds, and minutes) I doubt it will happen. We can put a cap on how long the sliding window interval can get, e.g., it has to be under an hour.

Reply:
Remember that all the slices within the ring buffer for the sliding window are also free running, not delta. This makes our snapshot faster as we're just doing load + store. Previous implementation required calculating the delta during snapshot. The tradeoff is that now snapshotting is much cheaper and no longer shows as hot in perf top. The downside is that for reporting, we need to do that work. This tradeoff seems reasonable for write-heavy use-cases (which I expect to be the typical pattern).

If we want we can drop this down to 32bit for all counters and just know that we need to do a narrowing conversion when we read external 64bit counters and store them into the live histogram. We can also try again to get 32bit BPF counters to work... but I had difficulty with that in the initial implementations (for reasons I can't recall clearly). 

Putting an upper-bound on the interval length makes sense - we have a lower bound of 1ms currently because on my macbook, 1us is too fine for snapshotting to complete on increment once we fall behind for the atomic case.

Reply:
Upper bound of 1 hour interval is implemented. 

Thinking about 32 bit vs 64 bit - I think we're probably safe with 32bit with the upper bound in place. At 1MHz (and assuming all counts go to the same bucket) 32bit covers > 1hour. At 1GHz increment rate (again in same bucket) it's ~4.29s between rollovers. This seems acceptable.

Question is - do we want to do that for the live histogram too? Do we make all the histograms 32bit? Do we have both 32 and 64 bit versions of the non-sliding window histograms?

Reply:
I'm beginning to think that for simplicity we just use 64-bit counters everywhere until we actually encounter a situation where the memory usage is problematic.

Reply:
I'm fine with that. We can cut an issue on github to remind ourselves and/or allow people to bump it in the future.

Reply:
That sounds like a good plan for now.